spec/mpt-proof_feat-refactor.md

```diff
@@ -0,0 +1,375 @@
+
+
+### SegmentTypes
+
+According to the [zkTrie spec], the leafnode hash is calculated by the formula
+`leafNodeHash = h(h(1, nodeKey), valueHash)`
+
+#### Account SegmentTypes
+
+For account, the formula for `valueHash` is
+![Account_ValueHash](https://i.imgur.com/6vahp0j.png)
+
+So we introduce account `SegmentType` as shown in the following figure
+
+![SegmentType_Account](https://hackmd.io/_uploads/S1Xew1L42.png)
+
+#### Storage SegmentTypes
+For storage, the formula for `valueHash` is
+`valueHash = h(storageValue[0:16], storageValue[16:32])`, so we have the following figure
+
+![SegmentType_Storage](https://hackmd.io/_uploads/rklHPNX4n.png)
+
+
+
+#### Expanding the trie leaf via SegmentTypes during circuit witness generation
+
+The above expansions are beyond what the original trie defined in [zkTrie spec] can have, where the expansion of the trie will only be up to leafnodes that are classified into `AccountTrie` or `StorageTrie`, i.e.  trie-type segments. The leaf-type segments (non-trie segments) are witnesses generated <i>inside</i> the circuit (parsed from `SMTTrace`), and the trie-like structure shown above are just imaginary/virtual which only lives in the circuit. So such common constraints as key is 0 and depth is 0 for non-trie segments must be enforced in the circuit constraint system. 
+
+The `PathType` used inside the circuit is also corresponding to this expanded trie including leaf-type segments (non-trie segments). For example, in the case of insert to fill operation, `PathType` changes from `Common` to `ExtensionNew` at the empty node that is filled by a new leaf, and the path continues to the specific account/storage field. So constraints for `PathType::ExtensionNew(Old)` must discuss separately for trie and non-trie segment cases.
+
+
+### Constraints 
+
+#### Shared Constraints
+for every row:
+- When `SegmentType==AccountTrie` or `StorageTrie` (`is_trie` is true):
+    - `key` at `depth` matches `direction`
+    - `depth_curr==depth_prev+1`
+    - when `PathType::Common`, `other_key` at `depth` matches `direction`
+    
+- When `SegmentType` is any other (non-trie) type (`is_trie` is false):
+    - `key==0`
+    - `depth==0` 
+
+- upper 128 bits of account/storage value is 16 bytes 
+
+
+#### PathType::Common
+on rows with `PathType::Common`:
+- for the old common path, `old_hash_previous==poisedon_hash(old_hash_left_child, old_hash_right_child)`
+- for the new common path, `new_hash_previous==poisedon_hash(new_hash_left_child, new_hash_right_child)`
+- for type 2 non-existence proof, common path includes an empty node as leafnode. So we provide constraints for this case
+    - if next row has `PathType::ExtensionNew` and `SegmentType::AccountTrie0`, then current row has `old_hash==0`
+    - if next row has `PathType::ExtensionOld` and `SegmentType::AccountTrie0`, then `new_hash==0`
+
+
+#### PathType::ExtensionNew
```

Comment:
Seems missing `PathType::ExtensionOld` (as constraint `new_value==0` and so on).